### PR TITLE
Add generic-array and vecarray support

### PR DESCRIPTION
**Pull request by sorpaas**
This adds generic-array and vecarray support in scale-codec.

### CODE REVIEW

File: src/codec.rs [hunk 1]
```diff
@@ -912,6 +912,51 @@ macro_rules! impl_non_endians {
 impl_endians!(u16, u32, u64, u128, i16, i32, i64, i128);
 impl_non_endians!(u8 {IS_U8}, i8, bool);
 
+#[cfg(feature = "generic-array")]
+impl<T: Encode, L: generic_array::ArrayLength<T>> Encode for generic_array::GenericArray<T, L> {
```
Comment (bkchr) [inline src/codec.rs]: Could you please move this into separate files? Like bitvec.

File: src/generic_array.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	fn encode_to<W: Output>(&self, dest: &mut W) {
+		for item in self.iter() {
+			item.encode_to(dest);
+		}
+	}
+}
+
+impl<T: Decode, L: generic_array::ArrayLength<T>> Decode for generic_array::GenericArray<T, L> {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		let mut r = Vec::new();
```
Comment (thiolliere) [inline src/generic_array.rs]: wouldn't it be better to have with_capacity here ? it doesn't harm as this capacity is on the type itself, nobody can craft it with wrong input.

File: src/lib.rs [hunk 3]
```diff
@@ -242,6 +242,10 @@ mod joiner;
 mod keyedvec;
 #[cfg(feature = "bit-vec")]
 mod bit_vec;
+#[cfg(feature = "generic-array")]
```
Comment (thiolliere) [inline src/lib.rs]: "generic-array" is not a feature in Cargo.toml

File: src/vec_array.rs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+	fn encode_to<W: Output>(&self, dest: &mut W) {
+		for item in self.iter() {
+			item.encode_to(dest);
+		}
+	}
+}
+
+impl<T: Decode, L: typenum::Unsigned> Decode for vecarray::VecArray<T, L> {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		let mut r = Vec::new();
```
Comment (thiolliere) [inline src/vec_array.rs]: wouldn't it be better to have with_capacity here ? it doesn't harm as this capacity is on the type itself, nobody can craft it with wrong input.

**Comment by thiolliere** [comment]
also add those features in CI

File: Cargo.toml [hunk 5]
```diff
@@ -14,6 +14,9 @@ serde = { version = "1.0", optional = true }
 parity-scale-codec-derive = { path = "derive", version = "1.0", default-features = false, optional = true }
 bitvec = { version = "0.14.0", default-features = false, features = ["alloc"], optional = true }
 byte-slice-cast = { version = "0.3.1", optional = true }
+generic-array = { version = "0.12", optional = true }
+vecarray = { version = "0.1", optional = true }
+typenum = { version = "1.10", optional = true }
```
Comment (thiolliere) [inline Cargo.toml]: do we need also to add `default-features = false` ? same for byte-slice-cast actually, is this needed

File: src/lib.rs (hunk 3 above)
Comment (bkchr) [reply src/lib.rs]: Dependencies are automatically added as features in Cargo.

File: Cargo.toml (hunk 5 above)
Comment (sorpaas) [reply Cargo.toml]: Fixed! `generic-array` and `typenum` does not need `std` feature.

Comment (sorpaas) [reply Cargo.toml]: `byte-slice-cast` pulls in `std` by default and does not support `no_std`. So when we're building `bitvec` it just won't work for `no_std` environments.

**Comment by thiolliere** [comment]
Looks good, do we want to implement optimisation for `T=u8` similarly to Encode of `[u8]` and Decode of `Vec<u8>` ?

File: src/generic_array.rs [hunk 6]
```diff
@@ -0,0 +1,25 @@
+use crate::{Encode, Decode, Error, Output, Input};
```
Comment (bkchr) [inline src/generic_array.rs]: Missing copyright and no tests.

File: src/vec_array.rs [hunk 7]
```diff
@@ -0,0 +1,22 @@
+use crate::{Encode, Decode, Error, Output, Input};
```
Comment (bkchr) [inline src/vec_array.rs]: Same.

File: src/generic_array.rs [hunk 8]
```diff
@@ -0,0 +1,64 @@
+// Copyright 2017, 2018 Parity Technologies
```
Comment (bkchr) [inline src/generic_array.rs]:
```suggestion
// Copyright 2019 Parity Technologies
```

File: src/vec_array.rs [hunk 9]
```diff
@@ -0,0 +1,61 @@
+// Copyright 2017, 2018 Parity Technologies
```
Comment (bkchr) [inline src/vec_array.rs]:
```suggestion
// Copyright 2019 Parity Technologies
```

**Pull request merged**